mlos_bench/mlos_bench/services/remote/ssh/ssh_service.py

```diff
@@ -0,0 +1,393 @@
+        self._connection_id: str = SshClient._CONNECTION_PENDING
+        self._connection: Optional[SSHClientConnection] = None
+        self._conn_event: CoroEvent = CoroEvent()
+        super().__init__(*args, **kwargs)
+
+    def __repr__(self) -> str:
+        return self._connection_id
+
+    def __str__(self) -> str:
+        return self.__repr__()
+
+    def __hash__(self) -> int:
+        return hash(self._connection_id)
+
+    @staticmethod
+    def id_from_connection(connection: SSHClientConnection) -> str:
+        """Gets a unique id repr for the connection."""
+        return f"{connection._username}@{connection._host}:{connection._port}"    # pylint: disable=protected-access
+
+    @staticmethod
+    def id_from_params(connect_params: dict) -> str:
+        """Gets a unique id repr for the connection."""
+        return f"{connect_params.get('username')}@{connect_params['host']}:{connect_params.get('port')}"
+
+    def connection_made(self, conn: SSHClientConnection) -> None:
+        """
+        Override hook provided by asyncssh.SSHClient.
+
+        Changes the connection_id from _CONNECTION_PENDING to a unique id repr.
+        """
+        self._conn_event.clear()
+        _LOG.debug("%s: Connection made by %s: %s", current_thread().name, conn._options.env, conn) \
+            # pylint: disable=protected-access
+        self._connection_id = SshClient.id_from_connection(conn)
+        self._connection = conn
+        self._conn_event.set()
+        return super().connection_made(conn)
+
+    def connection_lost(self, exc: Optional[Exception]) -> None:
+        self._conn_event.clear()
+        _LOG.debug("%s: %s", current_thread().name, "connection_lost")
+        if exc is None:
+            _LOG.debug("%s: gracefully disconnected ssh from %s: %s", current_thread().name, self._connection_id, exc)
+        else:
+            _LOG.debug("%s: ssh connection lost on %s: %s", current_thread().name, self._connection_id, exc)
+        self._connection_id = SshClient._CONNECTION_LOST
+        self._connection = None
+        self._conn_event.set()
+        return super().connection_lost(exc)
+
+    async def connection(self) -> Optional[SSHClientConnection]:
+        """
+        Waits for and returns the SSHClientConnection to be established or lost.
+        """
+        _LOG.debug("%s: Waiting for connection to be available.", current_thread().name)
+        await self._conn_event.wait()
+        _LOG.debug("%s: Connection available for %s", current_thread().name, self._connection_id)
+        return self._connection
+
+
+class SshClientCache:
+    """
+    Manages a cache of SshClient connections.
+    Note: Only one per event loop thread supported.
+    See additional details in SshService comments.
+    """
+
+    def __init__(self) -> None:
+        self._cache: Dict[str, Tuple[SSHClientConnection, SshClient]] = {}
+        self._cache_lock = CoroLock()
+
+    def __del__(self) -> None:
+        self.cleanup()
+
```

Comment:
I looked at that briefly, but it gets pretty complicated for our Services due to the way that mixins work.  We'd need to do something like a register enter/exit callbacks for each Service and then order may matter and ...

I can keep digging into that rock if you want, but in my testing so far this seems to work OK.

The only time it really matters to do the cleanup is on program exit anyways since all we need to do is stop a background thread.  The main reason to add the explicit `close()` call was for testing, and that's only to make the call to the cleanup code explicit so I could test it.

Thoughts?

Reply:
@motus did you have further thoughts on this thread?  Punt for now and revisit later or try and introduce contextmanager support for all Services first?